File: Wanda/ROS/src/hardware_control/nodes/camera_image_node.py
#!/usr/bin/env python3
# Based on code from:
# - https://automaticaddison.com
# - https://pyimagesearch.com/2015/03/30/accessing-the-raspberry-pi-camera-with-opencv-and-python/

# Import the necessary libraries
import cv2
from cv_bridge import CvBridge # Package to convert between ROS and OpenCV Images
import numpy as np
import rospy
from sensor_msgs.msg import Image
from sensor_msgs.msg import CameraInfo

# Custom libraries
from system_globals import frames
from topic_messages import topics

#================================================
# Get global parameters
IS_WANDA = rospy.get_param( 'IS_WANDA', default=False )

frame = frames.frames()
top = topics.topics()

#==============================================================================
class camera_image_node():
    def __init__(self):
        # Tells rospy the name of the node.
        # Anonymous = True makes sure the node has a unique name. Random
        # numbers are added to the end of the name.
        rospy.init_node('camera_image_node', anonymous=False)

        # Control max rate of image capture, Hz
        self.frame_rate = 10 # Hz # ToDo: make this a parameter
        self.rate = rospy.Rate( self.frame_rate )

        self.image_encoding = "bgr8" # ToDo: make this a parameter

        # Set up camera info parameters
        self.cinf = CameraInfo()
        self.cinf.header.stamp = rospy.Time.now()
        self.cinf.header.frame_id = frame.camera
        self.cinf.height = rospy.get_param( '/camera/image_height', default=720 )
        self.cinf.width = rospy.get_param( '/camera/image_width', default=1280 )
        self.cinf.distortion_model = rospy.get_param( '/camera/distortion_model', default='plumb_bob' )
        self.cinf.D = rospy.get_param( '/camera/distortion_coefficients/data', default=np.zeros((1,5)) )
        self.cinf.K = rospy.get_param( '/camera/camera_matrix/data', default=np.zeros((1,9)) )
        self.cinf.R = rospy.get_param( '/camera/rectification_matrix/data', default=np.zeros((1,9)) )
        self.cinf.P = rospy.get_param( '/camera/projection_matrix/data', default=np.zeros((1,12)) )
        self.cinf.binning_x = 0
        self.cinf.binning_y = 0
        self.info_pub = rospy.Publisher( top.camera_info, CameraInfo, queue_size=1 )

        # Create object to capture image
        self.cam = cv2.VideoCapture( 0 ) # The argument '0' gets the default webcam

        # Used to convert between ROS and OpenCV images
        self.br = CvBridge()

        # Node is publishing to the video_frames topic using the message type Image
        self.pub = rospy.Publisher( top.camera_image_raw, Image, queue_size=1 )

        self.publish_video()

    #================================================
    def publish_video( self ):
        ''' Main while loop to publish video data'''

        while not rospy.is_shutdown():
            # Record image
            image_captured, image = self.cam.read()

            if image_captured == True:

                if IS_WANDA:
                    # The camera is mounted upside down in the seeker
                    # Flip the image around
                    image = cv2.rotate( image, cv2.ROTATE_180 )

                ( rows, cols, channels ) = image.shape

                # Publish the image
                header_stamp = rospy.Time.now()
                image_message = self.br.cv2_to_imgmsg( image, encoding=self.image_encoding )
                image_message.header.stamp = header_stamp
                image_message.header.frame_id = frame.camera
                image_message.encoding = self.image_encoding
                image_message.height = rows # Include size of the image
                image_message.width = cols
                self.pub.publish( image_message )

                # Send camera info out as a topic
                self.cinf.header.stamp = header_stamp
                self.cinf.height = rows
                self.cinf.width = cols
                self.info_pub.publish( self.cinf )

            else:
                rospy.logwarn( f"Could not capture image frame at time = {rospy.Time.now()}" )

            # Sleep just enough to maintain the desired rate
            self.rate.sleep()

#==============================================================================
if __name__ == '__main__':
    try:
        cam_node = camera_image_node()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass

File: Wanda/ROS/src/hardware_control/src/camera_image_node.cpp

#include <ros/ros.h>
#include <cv_bridge/cv_bridge.h>
#include <image_transport/image_transport.h>
#include "opencv2/opencv.hpp"
#include <sensor_msgs/CameraInfo.h>
#include <string>
#include <vector>

// Custom classes
#include "hardware_control/camera_image.h"
#include "system_globals/frames.h"
#include "topic_messages/topics.h"

// SAS: for debugging
#include <iostream>

//===============================================
// Constructor
camera_image::camera_image( ros::NodeHandle* nh_in ):
nh( *nh_in ),
node_name( ros::this_node::getName() )
{
    processUserParams();

    image_transport::ImageTransport image_transport( nh );
    pub = image_transport.advertise( image_topic, 1 );
}

//=============================================================================
void camera_image::processUserParams()
{
    // Figure out if we want to flip the image over, if the camera is mounted upside down
    nh.param<bool>( "IS_WANDA", is_wanda, false );
    nh.param<bool>( node_name + "/rotate_image", rotate_image, false );

    // Grab namespace to publish everything out to
    nh.param<std::string>( node_name + "/namespace", output_namespace, "camera" );

    // Process if users want to rename the output topic
    if ( !nh.param<std::string>( node_name + "/image_topic", image_topic, "image_raw" ) ){
        // Use the topics class to grab global class names
        Topics top( nh );
        image_topic = top.camera_image_raw();
    }
    else{
        // Prepend the desired namespace to the topic
        image_topic = output_namespace + "/" + image_topic;
    }

    // Allow users to set the TF frame of the camera
    if ( !nh.param<std::string>( node_name + "/camera_frame", camera_frame, "seeker/camera" ) ){
        // Optionally, we can look for frame parameters under the "frames" namespace
        Frames frames( nh );
        camera_frame = frames.camera();
        nh.param<std::string>(  "frames/camera", camera_frame, "seeker/camera" );
    }

    // Allow users to set the frame rate
    nh.param<int>( node_name + "/frame_rate", video_hz, 10 );

    // Allow users to choose which camera hardware to use
    nh.param<int>( node_name + "/camera_source", camera_source, 0 );

    // We can optionally publish CameraInfo along with the image
    nh.param<bool>( node_name + "/publish_camera_info", publish_camera_info, false );
    if ( publish_camera_info ){
        // Grab the camera info parameters
        getCameraInfo();
    }
}

//=============================================================================
void camera_image::getCameraInfo()
{
    // Grab the namespace that all the camera parameters are stored under
    nh.param<std::string>(node_name + "/camera_info_namespace", camera_info_namespace, "camera");

    // Grab image shape
    int width, height;
    nh.param<int>(camera_info_namespace + "/image_width", width, 1280 );
    nh.param<int>(camera_info_namespace + "/image_height", height, 720 );
    caminfo.width = width;
    caminfo.height = height;

    // Grab camera matrix
    std::vector<float> camera_matrix;
    if ( !nh.getParam(camera_info_namespace + "/camera_matrix/data", camera_matrix) ) {
        ROS_WARN( "Could not find camera_matrix/data" );
    }
    else {
        for (unsigned int indx=0; indx < camera_matrix.size(); indx++){
            caminfo.K[indx] = camera_matrix[indx];
        }
    }

    // Grab distortion model type
    std::string distortion_model;
    nh.param<std::string>(camera_info_namespace + "/distortion_model", distortion_model, "plumb_bob");
    caminfo.distortion_model = distortion_model;

    // Grab distortion coeffiicents
    std::vector<float> distortion_coefficients;
    if ( !nh.getParam(camera_info_namespace + "/distortion_coefficients/data", distortion_coefficients) ) {
        ROS_WARN( "Could not find distortion_coefficients/data");
    }
    else {
        for ( unsigned int indx=0; indx < distortion_coefficients.size(); indx++){
            caminfo.P[indx] = distortion_coefficients[indx];
        }
    }

    // Grab rectification matrix
    std::vector<float> rectification_matrix;
    if ( !nh.getParam(camera_info_namespace + "/rectification_matrix/data", rectification_matrix) ) {
        ROS_WARN( "Could not find rectification_matrix/data" );
    }
    else {
        for ( unsigned int indx=0; indx < rectification_matrix.size(); indx++){
            caminfo.R[indx] = rectification_matrix[indx];
        }
    }

    // Grab projection matrix
    std::vector<float> projection_matrix;
    if ( !nh.getParam(camera_info_namespace + "/projection_matrix/data", projection_matrix) ) {
        ROS_WARN( "Could not find projection_matrix/data" );
    }
    else {
        for ( unsigned int indx=0; indx < projection_matrix.size(); indx++){
            caminfo.P[indx] = projection_matrix[indx];
        }
    }

    // Grab the topic name to publish cameraInfo to, under the output_namespace
    if ( !nh.param<std::string>(node_name + "/camera_info_topic", camera_info_topic, "camera_info") ){
        // Use default Topics class
        Topics top( nh );
        camera_info_topic = top.camera_info();
    }
    else{
        // Prepend the desired namespace
        camera_info_topic = output_namespace + "/" + camera_info_topic;
    }

    // Set up the publisher for this data
    pub_cam_info = nh.advertise<sensor_msgs::CameraInfo>(camera_info_topic, 1);
}

//=============================================================================
void camera_image::publish_video(){

    // Set up the camera
    cv::VideoCapture cam( camera_source );

    // Make sure we actually opened the camera.
    // If the camera doesn't work, try a few more times
    int fail_count = 0;
    while (!cam.isOpened() ){
        fail_count++;
        if ( fail_count > 10 ){
            ROS_ERROR( "Could not open video camera source %d. Quitting %s", camera_source, node_name.c_str() );
            return;
        }
    }

    // Set up loop rate
    ros::Rate loop_rate( video_hz );

    // Declare the image from the camera
    cv::Mat cv_image;

    // Declare the image message
    sensor_msgs::ImagePtr msg;

    // Loop forever to capture video feed
    while ( nh.ok() ){

        // Grab image from the camera
        cam >> cv_image;

        // If we actually got video feed, send it out
        if ( !cv_image.empty() ){

            // If we need to spin the image over, do it now
            if ( is_wanda || rotate_image ){
                cv::rotate( cv_image, cv_image, cv::ROTATE_180 );
            }

            // Package the image up into a message using cv_bridge
            msg = cv_bridge::CvImage(
                std_msgs::Header(),
                "bgr8",
                cv_image).toImageMsg();

            msg->header.stamp = ros::Time::now();
            msg->header.frame_id = camera_frame;

            // Publish the image
            pub.publish( msg );

            // Send out cameraInfo along with the image
            if ( publish_camera_info ){
                caminfo.header.stamp == ros::Time::now();
                pub_cam_info.publish( caminfo );
            }
        }

        // Wait until it's time to grab the next frame
        ros::spinOnce();
        loop_rate.sleep();
    }

    // Stop all threads
    cam.release();
    ros::shutdown();
}

//=============================================================================
int main( int argc, char** argv){

    ros::init(argc, argv, "camera_image_node");
    ros::NodeHandle node_handler;

    camera_image cam_node( &node_handler );
    cam_node.publish_video();
    ros::spin();

    return 0;
}
